tools/scanner/cmd/scanner.go

```diff
@@ -0,0 +1,136 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/osv/tools/scanner/internal/osv"
+	"github.com/google/osv/tools/scanner/internal/sbom"
+)
+
+func usage() {
+	fmt.Fprintf(flag.CommandLine.Output(), "%s [flags] dir1 dir2...\n", os.Args[0])
+	flag.PrintDefaults()
+}
+
+func scan(arg string) error {
+	info, err := os.Stat(arg)
+	if err != nil {
+		return err
+	}
+
+	if info.IsDir() {
+		return scanDir(arg)
+	}
+
+	// Try guessing file type and scanning the SBOM.
+	log.Printf("Scanning file %s\n", arg)
+	file, err := os.Open(arg)
+	if err != nil {
+		return err
+	}
+
+	scanIdentifier := func(id sbom.Identifier) error {
+		resp, err := osv.MakePURLRequest(id.PURL)
+		if err != nil {
+			return err
+		}
+		printResults(id.PURL, resp)
+		return nil
+	}
+
+	for _, provider := range sbom.Providers {
+		file.Seek(0, os.SEEK_SET)
+		err := provider.GetPackages(file, scanIdentifier)
+		if err == nil {
+			// Found the right format.
+			log.Printf("Scanned %s SBOM", provider.Name())
+			return nil
+		}
+
+		if errors.Is(err, sbom.InvalidFormat) {
+			continue
+		}
+
+		return err
+	}
+	return nil
+}
+
+func scanDir(dir string) error {
+	log.Printf("Scanning dir %s\n", dir)
+	return filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			log.Printf("Failed to walk %s: %v", path, err)
+			return err
+		}
+
+		if info.IsDir() && info.Name() == ".git" {
+			err = doScan(filepath.Dir(path))
+			if err != nil {
+				log.Printf("scan failed for %s: %v\n", path, err)
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+func getCommitSHA(repoDir string) (string, error) {
+	cmd := exec.Command("git", "-C", repoDir, "rev-parse", "HEAD")
+	var out bytes.Buffer
+	cmd.Stdout = &out
+
+	err := cmd.Run()
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(out.String()), nil
+}
+
+func doScan(repoDir string) error {
```

Comment:
How does `doScan()` differ to `scan()`? The names imply they are the same - is there a better name for this one?

Reply:
Good point. renamed `doScan` to `scanGit`.